// @repo/core/dist/schemas/Auth.schema.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.UpdateUserSchema = exports.CreateUserSchema = exports.RegisterSchema = exports.LoginSchema = void 0;
const zod_1 = require("zod");
exports.LoginSchema = zod_1.z.object({
    email: zod_1.z.string().email({ message: 'Por favor, introduce un email válido.' }),
    pass: zod_1.z.string().min(1, { message: 'La contraseña es obligatoria.' }),
});
exports.RegisterSchema = zod_1.z.object({
    name: zod_1.z.string().min(2, { message: 'El nombre debe tener al menos 2 caracteres.' }),
    email: zod_1.z.string().email({ message: 'Por favor, introduce un email válido.' }),
    password: zod_1.z.string().min(8, { message: 'La contraseña debe tener al menos 8 caracteres.' }),
    confirmPassword: zod_1.z.string().min(8, { message: 'La confirmación de la contraseña es obligatoria.' }),
}).refine((data) => data.password === data.confirmPassword, {
    message: 'Las contraseñas no coinciden.',
    path: ['confirmPassword'],
});
exports.CreateUserSchema = zod_1.z.object({
    name: zod_1.z.string().min(2),
    email: zod_1.z.string().email(),
    password: zod_1.z.string().min(8),
});
exports.UpdateUserSchema = zod_1.z.object({
    name: zod_1.z.string().min(2).optional(),
    email: zod_1.z.string().email().optional(),
    password: zod_1.z.string().min(8).optional(),
    currentPassword: zod_1.z.string().optional(),
    updateCustomerInfo: zod_1.z.boolean().optional().default(false),
}).refine(data => {
    // Si se proporciona password, currentPassword es obligatorio
    return !data.password || (data.password && data.currentPassword);
}, {
    message: 'Se requiere la contraseña actual para cambiar la contraseña',
    path: ['currentPassword'],
});
